File: coding-challenges/sources/daily-coding-problem/Day 19/Day19.kt
fun main() {
    // Test 1
    val matrix1 = arrayOf(
        intArrayOf(1, 2, 3),
        intArrayOf(4, 5, 6),
        intArrayOf(7, 8, 9)
    )
    val expected1 = 13
    val result1 = buildHouses(matrix1)
    println(result1 == expected1)

    // Test 2
    val matrix2 = arrayOf(
        intArrayOf(10, 11, 12),
        intArrayOf(13, 14, 15),
        intArrayOf(16, 17, 18)
    )
    val expected2 = 40
    val result2 = buildHouses(matrix2)
    println(result1 == expected1)

    // Test 3
    val matrix3 = arrayOf(
        intArrayOf(1, 2, 3),
        intArrayOf(4, 5, 6)
    )
    val expected3 = 6
    val result3 = buildHouses(matrix3)
    println(result3 == expected3)

    // Test 4
    val matrix4 = arrayOf(intArrayOf(1, 2, 3))
    val expected4 = 1
    val result4 = buildHouses(matrix4)
    println(result4 == expected4)
}

fun buildHouses(matrix: Array<IntArray>): Int {
    var lowestCost: Int = 0
    var lowestCostIndex: Int = -1
    var secondLowestCost: Int = 0
    
    for (row in matrix) {
        var newLowestCost: Int = Int.MAX_VALUE
        var newLowestCostIndex: Int = -1
        var newSecondLowestCost: Int = Int.MAX_VALUE
        
        for ((c, value) in row.withIndex()) {
            val prevLowestCost: Int = if (c == lowestCostIndex) secondLowestCost else lowestCost
            val cost: Int = prevLowestCost + value
            if (cost < newLowestCost) {
                newSecondLowestCost = newLowestCost
                newLowestCost = cost
                newLowestCostIndex = c
            } else if (cost < newSecondLowestCost) {
                newSecondLowestCost = cost
            }
        }
        
        lowestCost = newLowestCost
        lowestCostIndex = newLowestCostIndex
        secondLowestCost = newSecondLowestCost
    }
    
    return lowestCost
}

File: coding-challenges/sources/daily-coding-problem/Day 19/Day19.swift


func buildHouses(_ matrix: [[Int]]) -> Int {
    // Initialize variables
    var lowestCost = 0
    var lowestCostIndex = -1
    var secondLowestCost = 0

    // Iterate through rows of the matrix
    for row in matrix {
        // Initialize variables for new minimum cost, new lowest cost index, and new second lowest cost
        var newLowestCost = Int.max
        var newLowestCostIndex = -1
        var newSecondLowestCost = Int.max

        // Iterate through the columns of the current row
        for (c, val) in row.enumerated() {
            // Calculate the cost of painting the current house with the current color
            let prevLowestCost = (c == lowestCostIndex) ? secondLowestCost : lowestCost
            let cost = prevLowestCost + val

            // Update the minimum, second minimum, and index variables
            if cost < newLowestCost {
                newSecondLowestCost = newLowestCost
                newLowestCost = cost
                newLowestCostIndex = c
            } else if cost < newSecondLowestCost {
                newSecondLowestCost = cost
            }
        }

        // Update the main minimum, second minimum, and index variables
        lowestCost = newLowestCost
        lowestCostIndex = newLowestCostIndex
        secondLowestCost = newSecondLowestCost
    }

    // Return the minimum cost of painting the last house
    return lowestCost
}

import XCTest

final class Day19: XCTestCase {
    func test() {
        // Test 1
        let matrix1 = [
            [1, 2, 3],
            [4, 5, 6],
            [7, 8, 9]
        ]
        let expected1 = 13
        let result1 = buildHouses(matrix1)
        XCTAssertEqual(result1, expected1)

        // Test 2
        let matrix2 = [
            [10, 11, 12],
            [13, 14, 15],
            [16, 17, 18]
        ]
        let expected2 = 40
        let result2 = buildHouses(matrix2)
        XCTAssertEqual(result2, expected2)

        // Test 3
        let matrix3 = [
            [1, 2, 3],
            [4, 5, 6]
        ]
        let expected3 = 6
        let result3 = buildHouses(matrix3)
        XCTAssertEqual(result3, expected3)

        // Test 4
        let matrix4 = [[1, 2, 3]]
        let expected4 = 1
        let result4 = buildHouses(matrix4)
        XCTAssertEqual(result4, expected4)
    }

}


